File: pythonX/hard/h003_median_two_sorted_arrays.py
"""
H003: Median of Two Sorted Arrays

Summary:
Find median of two sorted arrays without merging.

Requirements:
Given two sorted arrays, find median efficiently.

Approach (Simplified):
Merge arrays using two pointers, then find median.
For optimal O(log(min(m,n))), use binary search (advanced).

Data Structures:
- List for merged array

Big O Analysis:
Time Complexity: O(m + n) for merge approach
Space Complexity: O(m + n)
"""


def find_median(arr1: list[int], arr2: list[int]) -> float:
    """
    TODO: Find median of two sorted arrays
    1. Merge arrays using two pointers
    2. Calculate median:
       - If odd length: middle element
       - If even length: average of two middle elements
    """
    pass


if __name__ == "__main__":
    # Test Cases
    
    print("Test 1: Odd total length")
    arr1 = [1, 3]
    arr2 = [2]
    print(f"Median of {arr1} and {arr2}: {find_median(arr1, arr2)}")  # Expected: 2.0
    
    print("\nTest 2: Even total length")
    arr3 = [1, 2]
    arr4 = [3, 4]
    print(f"Median of {arr3} and {arr4}: {find_median(arr3, arr4)}")  # Expected: 2.5
    
    print("\nTest 3: Same elements")
    arr5 = [0, 0]
    arr6 = [0, 0]
    print(f"Median of {arr5} and {arr6}: {find_median(arr5, arr6)}")  # Expected: 0.0
    
    print("\nTest 4: One empty array")
    arr7 = []
    arr8 = [1]
    print(f"Median of {arr7} and {arr8}: {find_median(arr7, arr8)}")  # Expected: 1.0
    
    print("\nTest 5: Other empty array")
    arr9 = [2]
    arr10 = []
    print(f"Median of {arr9} and {arr10}: {find_median(arr9, arr10)}")  # Expected: 2.0
    
    print("\nTest 6: Different lengths")
    arr11 = [1, 2, 3, 4, 5]
    arr12 = [6, 7, 8]
    print(f"Median of {arr11} and {arr12}: {find_median(arr11, arr12)}")  # Expected: 4.5

File: src/com/example/javas/hard/H003_MedianTwoSortedArrays.java
/*
HARD PROBLEM (Simplified for 25 min):
Find the median of two sorted arrays

REQUIREMENTS:
- Given two sorted arrays, find the median of the combined sorted array
- Do NOT merge the arrays explicitly (efficient solution required)
- Return median as double (average of two middle elements if even length)

APPROACH:
- Use binary search on the smaller array
- Partition both arrays such that left half has median
- Handle edge cases for empty partitions

TIME COMPLEXITY: O(log(min(m,n)))
SPACE COMPLEXITY: O(1)

NOTE: Simplified approach - merge and find median (easier for interview)
*/

package com.example.javas.hard;

public class H003_MedianTwoSortedArrays {
    
    public static double findMedian(int[] arr1, int[] arr2) {
        // TODO: Implement median finder (simplified merge approach)
        // 1. Create merged array of size arr1.length + arr2.length
        // 2. Use two pointers to merge both arrays in sorted order
        // 3. Find median:
        //    - If total length is odd: return middle element
        //    - If even: return average of two middle elements
        
        return 0.0;
    }
    
    // TEST CASES
    public static void main(String[] args) {
        int[] arr1 = {1, 3};
        int[] arr2 = {2};
        System.out.println("Median of [1,3] and [2]: " + findMedian(arr1, arr2));
        
        int[] arr3 = {1, 2};
        int[] arr4 = {3, 4};
        System.out.println("Median of [1,2] and [3,4]: " + findMedian(arr3, arr4));
        
        int[] arr5 = {0, 0};
        int[] arr6 = {0, 0};
        System.out.println("Median of [0,0] and [0,0]: " + findMedian(arr5, arr6));
        
        int[] arr7 = {};
        int[] arr8 = {1};
        System.out.println("Median of [] and [1]: " + findMedian(arr7, arr8));
        
        int[] arr9 = {2};
        int[] arr10 = {};
        System.out.println("Median of [2] and []: " + findMedian(arr9, arr10));
    }
}
